src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -423,50 +424,97 @@ private TypeSymbol NarrowVariableAccessType(TypeValidatorConfig config, Variable
             return targetType;
         }
 
+
         private TypeSymbol NarrowUnionType(TypeValidatorConfig config, SyntaxBase expression, TypeSymbol expressionType, UnionType targetType)
+            => expressionType switch
+            {
+                UnionType expressionUnion => NarrowUnionTypeForUnionExpressionType(config, expression, expressionUnion, targetType),
+                _ => NarrowUnionTypeForSingleExpressionType(config, expression, expressionType, targetType),
+            };
+
+        private TypeSymbol NarrowUnionTypeForUnionExpressionType(TypeValidatorConfig config, SyntaxBase expression, UnionType expressionType, UnionType targetType)
         {
             List<UnionTypeMemberViabilityInfo> candidacyEvaluations = new();
 
-            foreach (var candidateType in targetType.Members)
+            foreach (var candidateExpressionType in expressionType.Members)
             {
-                if (!AreTypesAssignable(expressionType, candidateType.Type))
+                if (!AreTypesAssignable(candidateExpressionType.Type, targetType))
                 {
-                    candidacyEvaluations.Add(new(candidateType, null, ImmutableArray<IDiagnostic>.Empty));
+                    candidacyEvaluations.Add(new(candidateExpressionType, null, ImmutableArray<IDiagnostic>.Empty));
                     continue;
                 }
 
                 var candidateDiagnostics = ToListDiagnosticWriter.Create();
                 var candidateCollector = new TypeValidator(typeManager, binder, candidateDiagnostics);
-                var narrowed = candidateCollector.NarrowType(config, expression, candidateType.Type);
+                var narrowed = candidateCollector.NarrowUnionTypeForSingleExpressionType(config, expression, candidateExpressionType.Type, targetType);
+                candidacyEvaluations.Add(new(candidateExpressionType, narrowed, candidateDiagnostics.GetDiagnostics()));
+            }
+
+            // report all informational and warning diagnostics
+            diagnosticWriter.WriteMultiple(candidacyEvaluations.SelectMany(e => e.Diagnostics.Where(d => d.Level != DiagnosticLevel.Error)));
 
-                // if we have a perfect match, skip checking the rest of the union members
-                if (!candidateDiagnostics.GetDiagnostics().Any())
+            var errors = candidacyEvaluations.SelectMany(c =>
+            {
+                var reportedErrors = c.Diagnostics.OfType<ErrorDiagnostic>();
```

Comment:
Note that it's possible to create regular `Diagnostic` of level `Error` - feels safer to invert the check you have on line 454